internal/core/app_exec.go

```diff
@@ -0,0 +1,257 @@
+package core
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes/any"
+	"github.com/hashicorp/go-argmapper"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/hcl/v2"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/hashicorp/waypoint-plugin-sdk/component"
+	"github.com/hashicorp/waypoint/internal/ceb"
+	"github.com/hashicorp/waypoint/internal/config"
+	pb "github.com/hashicorp/waypoint/internal/server/gen"
+)
+
+// Exec launches an exec plugin
+func (a *App) Exec(ctx context.Context, id string, d *pb.Deployment) error {
+	// We need to get the pushed artifact if it isn't loaded.
+	// TODO(evanphx) I don't understand why we do this. We're mimicing what the
+	// deploy destroy code does and they load this, so we do to?
+	var artifact *pb.PushedArtifact
+	if d.Preload != nil && d.Preload.Artifact != nil {
+		artifact = d.Preload.Artifact
+	}
+	if artifact == nil {
+		a.logger.Debug("querying artifact", "artifact_id", d.ArtifactId)
+		resp, err := a.client.GetPushedArtifact(ctx, &pb.GetPushedArtifactRequest{
+			Ref: &pb.Ref_Operation{
+				Target: &pb.Ref_Operation_Id{
+					Id: d.ArtifactId,
+				},
+			},
+		})
+		if status.Code(err) == codes.NotFound {
+			resp = nil
+			err = nil
+		}
+
+		if err != nil {
+			a.logger.Error("error querying artifact",
+				"artifact_id", d.ArtifactId,
+				"error", err)
+			return err
+		}
+
+		artifact = resp
+	}
+
+	// Add our build to our config
+	var evalCtx hcl.EvalContext
+	if err := evalCtxTemplateProto(&evalCtx, "artifact", artifact); err != nil {
+		a.logger.Warn("failed to prepare template variables, will not be available",
+			"err", err)
+	}
+
+	// Start the plugin
+	c, err := componentCreatorMap[component.PlatformType].Create(ctx, a, &evalCtx)
+	if err != nil {
+		a.logger.Error("error creating component in platform", "error", err)
+		return err
+	}
+	defer c.Close()
+
+	a.logger.Debug("spooling exec operation")
+
+	_, _, err = a.doOperation(ctx, a.logger.Named("exec"), &execOperation{
+		InstanceId: id,
+		Component:  c,
+		Deployment: d,
+		Client:     a.client,
+	})
+	return err
+}
+
+// execOperation is a simple barebones operation value to conform to the
```

Comment:
Do you need an operation? Or can you just call it directly?

Reply:
So I was wondering that and I'd love to chat about that with you. I was mimic'ing the other functions and they all seemed to use an operation.